### PR TITLE
dot/sync: handle block response justification 

### PR DESCRIPTION
**Pull request by noot**
## Changes

- handle justifications provided in block responses by setting latest finalized hash (will use grandpa to verify justification in a follow-up PR, wanted to get this in first since it allows the blocktree to be pruned, improving sync times)
- on node restart, if there was an "abnormal shutdown" ie the blocktree wasn't stored properly because the node panicked or was killed forcefully, then restore the state from the latest finalized hash
- update logging to log blocks imported every 5s (if you liked the previous logs, you can turn on `sync` package debug logs):
```
INFO[03-02|03:10:06] 🔨 finalized block                     pkg=sync number=1094832 hash=0xb7d4a3c8a51dd5e189c2aec82dffec97f91f2f09ce1736878aa985ec108eb41f caller=syncer.go:353
INFO[03-02|03:10:11] 🔗 imported blocks                     pkg=network from=1094832 to=1094918 hashes="[0xb7d4a3c8a51dd5e189c2aec82dffec97f91f2f09ce1736878aa985ec108eb41f ... 0xd7fd20dbe3646d2a52e3e82963ac00462cb6bfd95627a4f667903628f79000df]" caller=sync.go:275
INFO[03-02|03:10:11] 🚣 currently syncing                   pkg=network goal=6426917 average blocks/second=17.199 overall average=10.906 caller=sync.go:279
INFO[03-02|03:10:16] 🔗 imported blocks                     pkg=network from=1094918 to=1095008 hashes="[0xd7fd20dbe3646d2a52e3e82963ac00462cb6bfd95627a4f667903628f79000df ... 0x71192a673ef180eabf2df8c4baf8d0335e32cfbeda8f1bddaf6221aea0dc7774]" caller=sync.go:275
INFO[03-02|03:10:16] 🚣 currently syncing                   pkg=network goal=6426917 average blocks/second=17.988 overall average=11.087 caller=sync.go:279
```

## Tests

```
go test ./dot/state
go test ./dot/sync -short
```

as usual please try to run a kusama node and see that blocks are being marked as finalized

## Checklist

- [x] I have read [CODE_OF_CONDUCT](https://github.com/ChainSafe/gossamer/blob/development/.github/CODE_OF_CONDUCT.md) and [CONTRIBUTING](https://github.com/ChainSafe/gossamer/blob/development/.github/CONTRIBUTING.md) 
- [x] I have provided as much information as possible and necessary
- [x] I have reviewed my own pull request before requesting a review
- [x] All integration tests and required coverage checks are passing

## Issues

- related to #1363

### CODE REVIEW

File: dot/network/sync.go [hunk 1]
```diff
@@ -249,33 +251,46 @@ func (q *syncQueue) benchmark() {
 			return
 		}
 
-		head, err := q.s.blockState.BestBlockNumber()
+		before, err := q.s.blockState.BestBlockHeader()
 		if err != nil {
-			logger.Error("failed to get best block number", "error", err)
+			logger.Error("failed to get best block haeder", "error", err)
```
Comment (RyRy79261) [inline dot/network/sync.go:256]: Spelling error

File: dot/network/sync.go [hunk 2]
```diff
@@ -249,33 +251,46 @@ func (q *syncQueue) benchmark() {
...
-		q.benchmarker.begin(head.Uint64())
-		time.Sleep(time.Second * 10)
+		q.benchmarker.begin(before.Number.Uint64())
+		time.Sleep(time.Second * 5)
 
-		head, err = q.s.blockState.BestBlockNumber()
+		after, err := q.s.blockState.BestBlockHeader()
 		if err != nil {
-			logger.Error("failed to get best block number", "error", err)
+			logger.Error("failed to get best block haeder", "error", err)
```
Comment (RyRy79261) [inline dot/network/sync.go:269]: Spelling error

**Review by RyRy79261 — APPROVED** [review]
Just the two spelling error but the code looks good

File: dot/network/sync.go [hunk 3]
```diff
@@ -352,7 +367,7 @@ func (q *syncQueue) setBlockRequests(to peer.ID) {
 	newReqs = sortRequests(newReqs)
 	q.requests = newReqs
```
Comment (arijitAD) [inline dot/network/sync.go:368]: Acquire lock

**Review by arijitAD — COMMENTED** [review]
Wouldn't it be simpler to use channels directly to implement `syncQueue` so we don't have to worry about locks. The current code has a couple of data races.

File: dot/network/sync.go (hunk 3 above)
Comment (noot) [reply dot/network/sync.go:368]: there's currently no request lock

**Comment by noot** [comment]
@arijitAD what do you mean use channels directly? 

btw, I would rather not modify the sync queue in this PR since this PR was for handling justifications and it might cause other issues. I can address the issues in another PR

**Comment by arijitAD** [comment]
> @arijitAD what do you mean use channels directly?
> 
> btw, I would rather not modify the sync queue in this PR since this PR was for handling justifications and it might cause other issues. I can address the issues in another PR

Makes sense. We can definitely refactor this later.

```go
	requests  []*syncRequest // start block of message -> full message
	requestCh chan *syncRequest

	responses    []*types.BlockData
	responseCh   chan []*types.BlockData
```
Currently, we are adding the data to the `requests` slice and then moving it to the `requestCh` later. Can we directly add the `syncRequest` to the `requestCh`? That way we won't be needing locks since channels are threadsafe.

Also, we should have locks for `requests` since we are modifying it in `setBlockRequests` and reading it in `handleRequestQueue`.

**Comment by noot** [comment]
@arijitAD I see, I guess the reason I made it so that the requests are separate from the channel are in the case we need to modify the requests while they are queued / before they are sent out.  I think it could work with a buffered channel as well though, we would just need to know what's in the buffer somehow. I think the queue could definitely be improved so I'll address this in a follow up PR!

**Review by edwardmack — APPROVED** [review]

**Pull request merged**